### DISCUSSION TITLE
q-table custom header cell with normal quasar sorting

### DISCUSSION THREAD

**Discussion by ssuess** [quasarframework/quasar#16466] [General - Components / Directives / etc] [answered]
hi I have a custom table header cell that lets me update the contents of the table when an item is selected:
```
<template v-slot:header-cell-type="props">
    <th style="width: 120px;"><q-select
      v-model="selectedType"
      dense
      :options="typeOptions"
      label="Type"
      @update:model-value="handleTypeChange"
      style="margin-bottom:-5px;"
    /></th>
  </template>
  ```
  but I would ALSO like to have the standard sorting functionality and up/down arrow that I have with other sortable columns (to the right of my select dropdown). Is there a template somewhere for adding this functionality into the header slot like it is in default? I have tried several variations, and none of them work. Thanks!

**Comment by ssuess** [comment]
based on examining the code in the dev console, the closest I have come is this:
```
<template v-slot:header-cell-type="props">
    <th class="text-left sortable" style="width: 160px;"><q-select
      v-model="selectedType"
      dense
      :options="typeOptions"
      label="Type"
      @update:model-value="handleTypeChange"
      style="margin-bottom:-5px;width: 110px;"
    />
    <span style="width:25px;float:right;" class="q-icon q-table__sort-icon q-table__sort-icon--left" aria-hidden="true" role="presentation"><svg viewBox="0 0 256 512"><path d="M3.515 168.97l7.07 7.071c4.686 4.686 12.284 4.686 16.971 0L111 92.113V468c0 6.627 5.373 12 12 12h10c6.627 0 12-5.373 12-12V92.113l83.444 83.928c4.686 4.686 12.284 4.686 16.971 0l7.07-7.071c4.686-4.686 4.686-12.284 0-16.97l-116-116.485c-4.686-4.686-12.284-4.686-16.971 0L3.515 152c-4.687 4.686-4.687 12.284 0 16.97z"></path></svg></span></th>
  </template>
  ```
  Which will show the icon on hover, but no action of sorting happens when it is clicked on. Is there a sorting function of some sort I need to call, and if so what is it and how can I make it work like a normal sorted column?

**Comment by ssuess** [comment]
making progress, I can trigger the sort with `props.props.sort('type');` but not any of the effects of rotating/showing/hiding the arrow. Do I need to rebuild all of this myself or is there a way to leverage what quasar has inbuilt?

**Comment by ssuess** [comment, marked as answer]
This was a pain, but I got this working by recreating (and keeping track of) the sort state and using the quasar classes that I noticed were being set and removed with the other columns in my own functions.
```
<template v-slot:header-cell-type="props">
      <th nowrap :class="thClassType" 
        style="min-width: 140px;max-width:160px;"><q-select
      v-model="selectedType"
      dense
      :options="typeOptions"
      label="Type"
      @update:model-value="handleTypeChange"
      style="margin-bottom:-18px;min-width: 100px; max-width:calc(100% - 25px);"
    />
    <span @click="toggleSort('type'); props.props.sort('type');" style="width:20px;float:right;top:-10px;" class="q-icon q-table__sort-icon q-table__sort-icon--left" aria-hidden="true" role="presentation"><svg viewBox="0 0 256 512"><path d="M3.515 168.97l7.07 7.071c4.686 4.686 12.284 4.686 16.971 0L111 92.113V468c0 6.627 5.373 12 12 12h10c6.627 0 12-5.373 12-12V92.113l83.444 83.928c4.686 4.686 12.284 4.686 16.971 0l7.07-7.071c4.686-4.686 4.686-12.284 0-16.97l-116-116.485c-4.686-4.686-12.284-4.686-16.971 0L3.515 152c-4.687 4.686-4.687 12.284 0 16.97z"></path></svg></span>
  </th>
  </template>
  ```

**Reply by DetonatorB** [reply]
Although I generally dislike hacking it in like this, I believe we can improve on this already excellent answer.

`<template v-slot:header-cell-pd="props">
        <q-th :class="props.col.__thClass"><span @click="props.sort(props.col.name);">{{props.col.label}}<i :class="props.col.__iconClass" class="q-icon fas fa-arrow-up" aria-hidden="true" role="presentation"></i></span>
        </q-th>
      </template>`

This improves
-  flexibility (if you rename this name of the field, this keeps working)
- compatibility (a bit) because we use the classes assigned by QTable
- reduces complexity (you do not have to keep track of the sorting status)

Of course it would be even nicer of we could have a slot that does not do away with sorting or to be able to insert html in the header without losing sort control. 

Edit: Added span to include the label text in the @click handler so clicking in the label text also toggles the sorting
